### PR TITLE
Feat/pvp  handle season restart

### PR DESCRIPTION
**Pull request by Jcalcaldev**
This PR:

- Adds a counter so players know when a PvP ranked season restarts.
- Displays current season number.
- Adds an indicator on the "Rewards" tab so players know when they can claim rewards from being top players in a season.
- Notifies players the instant a season is restarted.

### CODE REVIEW

File: frontend/src/components/smart/PvPArena.vue [hunk 1]
```diff
@@ -170,6 +180,13 @@ export default {
       this.$emit('leaveMatchMaking');
     },
 
+    async updateSecondsBeforeSeason() {
+      this.currentRankedSeason = await this.contracts().PvpArena.methods.currentRankedSeason().call({ from: this.defaultAccount });
```
Comment (Chriserus) [inline frontend/src/components/smart/PvPArena.vue:184]: Don't use direct contract calls in vue files, keep it in store.ts, it applies to every call here

File: frontend/src/components/smart/PvPArena.vue [hunk 2]
```diff
@@ -372,10 +437,22 @@ export default {
     this.loading = false;
   },
 
+  async updated() {
+    await this.updateSecondsBeforeSeason();
+  },
+
+  destroyed() {
+    if (this.restartEventSubscription.id) {
```
Comment (Chriserus) [inline frontend/src/components/smart/PvPArena.vue:445]: Can restartEventSubscription be null here? It will throw an exception if it's possible

File: frontend/src/components/smart/PvPArenaPreparation.vue [hunk 3]
```diff
@@ -165,7 +165,18 @@
             <span>{{ tierTopRankers[2] && tierTopRankers[2].rank || 'N/A'}}</span>
           </li>
         </ul>
-        <!-- <a href="#" class="rankings">View all rankings</a> -->
+        <ul class="topPlayersList">
+          <li class="header">
+            <span>Current Season</span><span>Restarts In</span>
```
Comment (Chriserus) [inline frontend/src/components/smart/PvPArenaPreparation.vue:170]: After it's live someone will come back and clean all the translations! I'm watching...

File: frontend/src/components/smart/PvPArenaSummary.vue [hunk 4]
```diff
@@ -72,11 +72,20 @@
...
           </li>
         </ul>
-        <!-- <a href="/" class="rankings">View all rankings</a> -->
+        <ul class="topPlayersList">
+          <li class="header">
+            <span>Current Season</span><span>Restarts In</span>
+          </li>
+          <li>
+            <span>{{ currentRankedSeason }}</span>
+            <vue-countdown :time="secondsBeforeNextSeason * 1000" v-slot="{ days, hours, minutes, seconds }">
```
Comment (Chriserus) [inline frontend/src/components/smart/PvPArenaSummary.vue:81]: I see a duplication here, with stuff from PvPArenaPreparation, can we make a component out of it? (everything in < ul >)

File: frontend/src/components/smart/PvPNavBar.vue [hunk 5]
```diff
@@ -11,7 +11,13 @@
       </button>
       <button @click="setTab(2)" :class="tabNumber === 2 && 'active'">
         <img src="../../assets/crownIcon.svg" />
-        <span>Rewards</span>
+        <span>Rewards
+          <div class="rewardsIndicator" v-if="hasRewards">
+            <div>
+              <p>!</p>
```
Comment (Chriserus) [inline frontend/src/components/smart/PvPNavBar.vue:17]: I don't know how it will look like in browser, but I think in this case we can use an icon indicator and not literal "!"

File: frontend/src/views/PvP.vue [hunk 6]
```diff
@@ -34,12 +34,13 @@ export default {
...
       tab: 0,
-      isCharacterMatchMaking: false
+      isCharacterMatchMaking: false,
+      hasRewards: false,
     };
   },
 
   computed: {
-    ...mapState(['currentCharacterId']),
+    ...mapState(['currentCharacterId', 'contracts', 'defaultAccount']),
```
Comment (Chriserus) [inline frontend/src/views/PvP.vue:43]: Don't. Please extract to store.ts

**Review by Chriserus — CHANGES_REQUESTED** [review]

File: frontend/src/components/smart/PvPArena.vue (hunk 1 above)
Comment (RafaelVidaurre) [reply frontend/src/components/smart/PvPArena.vue:184]: @Chriserus I disagree with this, passing all calls through the store has no real value if no app state is affected. The main issue with the previous implementation of this feature was that it's whole state was passed through the store even though it was specific to this view.

Also that meant we had to manually reset the state on route changes.

File: frontend/src/components/smart/PvPArenaPreparation.vue (hunk 3 above)
Comment (Jcalcaldev) [reply frontend/src/components/smart/PvPArenaPreparation.vue:170]: Will be done

File: frontend/src/components/smart/PvPArenaInfo.vue [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+.title {
+  margin-bottom: 1rem;
+  color: #cec198;
+  font-size: 1.25rem;
+  font-family: 'Trajan';
+  line-height: 1.75rem;
+  padding: 0;
+}
+.tokenCard {
+display: flex;
```
Comment (Chriserus) [inline frontend/src/components/smart/PvPArenaInfo.vue:100]: No indent

File: frontend/src/components/smart/PvPArenaInfo.vue [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+  .number {
+    color: #ffffff;
+    font-size: 1.25rem;
+    line-height: 1.75rem;
+  }
+}
+}
+.topPlayersList,
+.characterAttrsList {
+display: flex;
```
Comment (Chriserus) [inline frontend/src/components/smart/PvPArenaInfo.vue:129]: Same here. Fixing this should be automatically done by IDE if configured

File: frontend/src/components/smart/PvPArena.vue (hunk 1 above)
Comment (Chriserus) [reply frontend/src/components/smart/PvPArena.vue:184]: "no real value", it's basic single responsibility principle. It's the same with any api calls, why do we create services for that? Won't it work directly from component? Ofc it will, but will be non-extendable and the SRP will be broken. That's why we have layers (components call api through services, web3 through store.ts) and at least some sort of architecture in our project. I know that single, huge store.ts is not ideal, I'm aware of that, but it's better to continue extracting calls there then to litter components with a responsibility that shouldn't be handled there.

Also, the solution you guys provided is error prone because it's not checking contracts or defaultAccount existence and they're not always available (see checks in every method in store).

Happy to discuss it further.

Comment (Chriserus) [reply frontend/src/components/smart/PvPArena.vue:184]: Also, "that meant we had to manually reset the state on route changes", I'm not talking about handling state of PvP in store. I just mean that when you have a simple getter like the line we're commenting under, it should be called this way:

this.currentRankedSeason = await this.currentRankedSeason(); 
In which, currentRankedSeason is an action imported from the store (ideally with an interface here showing what are the arguments and what is the output)

Comment (Chriserus) [reply frontend/src/components/smart/PvPArena.vue:184]: Example interface from another class: 
```
interface StoreMappedActions {
  fetchShields(shieldIds: (string | number)[]): Promise<void>;
  fetchWeapons(weaponIds: (string | number)[]): Promise<void>;
  fetchCharacters(characterIds: (string | number)[]): Promise<void>;
}
```

Comment (RafaelVidaurre) [reply frontend/src/components/smart/PvPArena.vue:184]: Replying to this now, sorry for the delay.

About SRP: I think this is not a case of SRP violation, one store doing every web3 call has multiple responsibilities. I think there's clear value in extracting reusable code, but these calls are literally only used for this component. I do think there's a strong case here in terms of this PR being inconsistent with the rest of the codebase though.

About "Checking for defaultAccount existence (should be)". I guess we assumed this values were guaranteed to be set by the time the route rendered. If not then yup, we have to do that, though I'd say we should consider this in the future as part of an architecture improvement, we shouldn't be repeating this logic on every view if we can extract it.

I think moving the calls themselves to the store is OK, if that's how we are doing it everywhere else, but the previous implementation had horrible issues with the **state** being global. If your request is just about moving the web3 calls into actions then we can definitely do that.

TL;DR;

* Happy to move web3 calls to actions for consistency
* I'm against moving view-specific state to the store (though I think I misunderstood, and you weren't suggesting this)
* I have to dissagree with this being a SRP issue in this feature, and I'd even argue it *is* a big issue on the store itself, as it should be subdivided into modules that do respect that principle (right now being at 4k+ lines), so if anything, that's where we should be looking to make improvements regarding separation of concerns first.

One last thing though. I think we should merge this PR and tackle the migration of web3 calls to actions. Because there's already multiple instances of these merged, and I think it is better if we tackle this on a separate PR

Comment (Chriserus) [reply frontend/src/components/smart/PvPArena.vue:184]: I'm aware of our overall architectural issues with having just one store, we're on the same page here. I didn't want you to move any pvp-specific states to the store, that's ok.

Just for the sake of consistency, please move all the web3 calls to store.ts, so we have it all in one place and not scattered in components.

I'm happy with doing it on a different PR in the future, I just want you to make sure it won't get lost due to new features or something, cleanup should take place in not-so-distant future.

As my other comments were formatting-related I will approve the PR and you can merge it in appropriate time.

@seiyria tagging you here, as we had a nice discussion with Rafael and I want it to be visible

**Review by Chriserus — APPROVED** [review]

File: frontend/src/components/smart/PvPArena.vue (hunk 1 above)
Comment (RafaelVidaurre) [reply frontend/src/components/smart/PvPArena.vue:184]: Thanks for the help @Chriserus , will do that
